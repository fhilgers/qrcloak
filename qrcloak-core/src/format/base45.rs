#[cfg(feature = "serde")]
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[cfg(feature = "json")]
use schemars::JsonSchema;

#[cfg(feature = "serde")]
pub struct Base45IfHumanReadable;

#[cfg(feature = "serde")]
impl Base45IfHumanReadable {
    pub fn serialize<S>(data: &bytes::Bytes, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        if serializer.is_human_readable() {
            base45::encode(data).serialize(serializer)
        } else {
            data.serialize(serializer)
        }
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<bytes::Bytes, D::Error>
    where
        D: Deserializer<'de>,
    {
        if deserializer.is_human_readable() {
            match base45::decode(<&str>::deserialize(deserializer)?) {
                Ok(buf) => Ok(buf.into()),
                Err(_) => Err(<D::Error as serde::de::Error>::invalid_value(
                    serde::de::Unexpected::Other("invalid base45 string"),
                    &"a valid base45 string",
                )),
            }
        } else {
            bytes::Bytes::deserialize(deserializer)
        }
    }
}

#[cfg(feature = "json")]
const BASE45_PATTERN: &str = "^[0-9A-Z\\s\\$%\\*\\+\\-\\.\\/:]*$";

#[cfg(feature = "json")]
impl JsonSchema for Base45IfHumanReadable {
    fn schema_name() -> String {
        "String".to_string()
    }

    fn json_schema(_: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
        schemars::schema::SchemaObject {
            instance_type: Some(schemars::schema::InstanceType::String.into()),
            format: Some("base45".into()),
            string: Some(
                schemars::schema::StringValidation {
                    pattern: Some(BASE45_PATTERN.into()),
                    ..Default::default()
                }
                .into(),
            ),
            ..Default::default()
        }
        .into()
    }

    fn is_referenceable() -> bool {
        false
    }
}
