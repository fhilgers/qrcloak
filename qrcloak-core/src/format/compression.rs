#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone, Default)]
pub enum CompressionSpec {
    #[default]
    NoCompression,
}

impl CompressionSpec {
    pub fn no_compression(&self) -> bool {
        match self {
            CompressionSpec::NoCompression => true,
        }
    }
}
