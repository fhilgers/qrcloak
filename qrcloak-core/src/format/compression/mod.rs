use bytes::Bytes;
use thiserror::Error;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

use super::CompletePayload;

#[derive(Debug, Clone, Default)]
pub enum Compression {
    #[default]
    NoCompression,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone, Default)]
pub enum CompressionSpec {
    #[default]
    NoCompression,
}

impl CompressionSpec {
    pub fn no_compression(&self) -> bool {
        match self {
            CompressionSpec::NoCompression => true,
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Default)]
pub enum Decompression {
    #[default]
    NoCompression,
}

#[derive(Debug, Error)]
pub enum CompressionError {}

impl Compression {
    pub fn process(&self, data: Bytes) -> Result<Bytes, CompressionError> {
        match self {
            Compression::NoCompression => Ok(data),
        }
    }

    pub fn spec(&self) -> CompressionSpec {
        match self {
            Compression::NoCompression => CompressionSpec::NoCompression,
        }
    }
}

#[derive(Debug, Error)]
pub enum DecompressionError {
    #[error("tried to decompress with {tried:?} but payload is compressed with {payload:?}")]
    SpecMismtach {
        payload: CompressionSpec,
        tried: Decompression,
    },
}

impl Decompression {
    pub fn process(&self, data: &mut CompletePayload) -> Result<(), DecompressionError> {
        match (self, &data.compression) {
            (Decompression::NoCompression, CompressionSpec::NoCompression) => Ok(()),
        }
    }
}
