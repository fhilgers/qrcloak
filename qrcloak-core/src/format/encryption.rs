#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone, Default)]
pub enum EncryptionSpec {
    #[default]
    NoEncryption,
    AgePassphrase,
    AgeKey,
}

impl EncryptionSpec {
    pub fn no_encryption(&self) -> bool {
        match self {
            EncryptionSpec::NoEncryption => true,
            _ => false,
        }
    }
}
