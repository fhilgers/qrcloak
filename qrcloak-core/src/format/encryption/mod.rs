use bytes::Bytes;
use thiserror::Error;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

use super::CompletePayload;

mod age_encryption;

pub use age_encryption::{AgeKeyDecryption, AgeKeyEncryption, AgePassphrase};

#[derive(Debug, Clone, Default)]
pub enum Encryption {
    #[default]
    NoEncryption,
    AgePasshprase(age_encryption::AgePassphrase),
    AgeKey(age_encryption::AgeKeyEncryption),
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone, Default)]
pub enum EncryptionSpec {
    #[default]
    NoEncryption,
    AgePassphrase,
    AgeKey,
}

impl EncryptionSpec {
    pub fn no_encryption(&self) -> bool {
        match self {
            EncryptionSpec::NoEncryption => true,
            _ => false,
        }
    }
}

#[derive(Debug, Clone, Default)]
pub enum Decryption {
    #[default]
    NoEncryption,
    AgePassphrase(age_encryption::AgePassphrase),
    AgeKey(age_encryption::AgeKeyDecryption),
}

#[derive(Debug, Error)]
pub enum EncryptionError {
    #[error(transparent)]
    Age(#[from] age::EncryptError),
}

#[derive(Debug, Error)]
pub enum DecryptionError {
    #[error("tried to decrypt with {tried:?} but payload is decrypted with {payload:?}")]
    SpecMismtach {
        payload: EncryptionSpec,
        tried: Decryption,
    },

    #[error(transparent)]
    Age(#[from] age::DecryptError),
}

impl Encryption {
    pub fn process(&self, data: Bytes) -> Result<Bytes, EncryptionError> {
        match self {
            Encryption::NoEncryption => Ok(data),
            Encryption::AgePasshprase(pw) => Ok(pw.encrypt(data)?),
            Encryption::AgeKey(key) => Ok(key.encrypt(data)?),
        }
    }

    pub fn spec(&self) -> EncryptionSpec {
        match self {
            Encryption::NoEncryption => EncryptionSpec::NoEncryption,
            Encryption::AgePasshprase(_) => EncryptionSpec::AgePassphrase,
            Encryption::AgeKey(_) => EncryptionSpec::AgeKey,
        }
    }
}

impl Decryption {
    pub fn process(&self, data: &mut CompletePayload) -> Result<(), DecryptionError> {
        match (self, &data.encryption) {
            (Decryption::NoEncryption, EncryptionSpec::NoEncryption) => Ok(()),
            (Decryption::AgePassphrase(pw), EncryptionSpec::AgePassphrase) => {
                data.data = pw.decrypt(data.data.clone())?;
                Ok(())
            }
            (Decryption::AgeKey(key), EncryptionSpec::AgeKey) => {
                data.data = key.decrypt(data.data.clone())?;
                Ok(())
            }
            (tried, payload) => Err(DecryptionError::SpecMismtach {
                payload: payload.clone(),
                tried: tried.clone(),
            }),
        }
    }
}
