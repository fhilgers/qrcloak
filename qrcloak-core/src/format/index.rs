#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

/// The index of a partial payload.
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Index {
    /// The id to match multiple partial payloads together
    pub(crate) id: u32,

    /// The index in the group
    pub(crate) index: u32,

    /// The total size of the group
    pub(crate) size: u32,
}

impl Index {
    /// Checks whether the index for the first
    /// element in the group
    pub fn is_head(&self) -> bool {
        self.index == 0
    }

    /// Checks whether the index for the remaining
    /// elements in the group
    pub fn is_tail(&self) -> bool {
        !self.is_head()
    }
}
