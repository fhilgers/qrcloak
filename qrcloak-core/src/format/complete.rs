use bytes::Bytes;

use super::{encryption::EncryptionSpec, CompressionSpec};

#[cfg(feature = "json")]
use crate::format::base45::Base45IfHumanReadable;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

/// A complete payload, meaning a payload that is not split
/// accross multiple partial ones.
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone)]
pub struct CompletePayload {
    /// The data of the payload.
    #[cfg_attr(feature = "serde", serde(with = "Base45IfHumanReadable"))]
    pub(crate) data: Bytes,

    /// The encryption to be used for the payload.
    #[cfg_attr(
        feature = "serde",
        serde(skip_serializing_if = "EncryptionSpec::no_encryption", default)
    )]
    pub(crate) encryption: EncryptionSpec,

    /// The compression to be used for the payload.
    #[cfg_attr(
        feature = "serde",
        serde(skip_serializing_if = "CompressionSpec::no_compression", default)
    )]
    pub(crate) compression: CompressionSpec,
}
