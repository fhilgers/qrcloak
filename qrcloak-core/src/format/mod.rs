mod base45;
mod complete;
mod compression;
mod encryption;
mod index;
mod partial;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[cfg_attr(feature = "serde", serde(untagged))]
#[derive(Debug, Clone)]
pub enum Payload {
    Complete(CompletePayload),
    Partial(PartialPayload),
}

impl From<CompletePayload> for Payload {
    fn from(payload: CompletePayload) -> Self {
        Self::Complete(payload)
    }
}

impl From<PartialPayload> for Payload {
    fn from(payload: PartialPayload) -> Self {
        Self::Partial(payload)
    }
}

impl TryFrom<Payload> for CompletePayload {
    type Error = Payload;

    fn try_from(payload: Payload) -> Result<Self, Self::Error> {
        match payload {
            Payload::Complete(payload) => Ok(payload),
            p => Err(p),
        }
    }
}

impl TryFrom<Payload> for PartialPayload {
    type Error = Payload;

    fn try_from(payload: Payload) -> Result<Self, Self::Error> {
        match payload {
            Payload::Partial(payload) => Ok(payload),
            p => Err(p),
        }
    }
}

pub use complete::CompletePayload;
pub use compression::CompressionSpec;
pub use encryption::EncryptionSpec;
pub use index::Index;
pub use partial::{PartialPayload, PartialPayloadHead, PartialPayloadTail};

#[cfg(all(test, feature = "json"))]
mod tests {
    use insta::assert_json_snapshot;
    use schemars::schema_for;

    use crate::format::Payload;

    #[test]
    fn validate_schema() {
        let schema = schema_for!(Payload);

        assert_json_snapshot!(schema);
    }
}
