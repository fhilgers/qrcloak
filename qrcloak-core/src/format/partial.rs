use bytes::Bytes;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

#[cfg(feature = "json")]
use schemars::JsonSchema;

use super::base45::Base45IfHumanReadable;
use super::{index::Index, CompletePayload};

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone)]
pub struct PartialPayloadHead {
    #[cfg_attr(feature = "serde", serde(flatten))]
    pub(crate) complete: CompletePayload,
    pub(crate) index: Index,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone)]
pub struct PartialPayloadTail {
    #[cfg_attr(feature = "serde", serde(with = "Base45IfHumanReadable"))]
    pub(crate) data: Bytes,
    pub(crate) index: Index,
}

#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[cfg_attr(feature = "json", derive(JsonSchema))]
#[derive(Debug, Clone)]
pub enum PartialPayload {
    Head(PartialPayloadHead),
    Tail(PartialPayloadTail),
}

impl PartialPayload {
    pub fn index(&self) -> Index {
        match self {
            PartialPayload::Head(head) => head.index,
            PartialPayload::Tail(tail) => tail.index,
        }
    }

    // TODO do not allow parsing of misconfigured
    pub fn is_misconfigured(&self) -> bool {
        match self {
            PartialPayload::Head(head) => head.index.is_tail(),
            PartialPayload::Tail(tail) => tail.index.is_head(),
        }
    }

    pub fn get_head(&self) -> Option<&PartialPayloadHead> {
        match self {
            PartialPayload::Head(head) => Some(head),
            _ => None,
        }
    }

    pub fn get_tail(&self) -> Option<&PartialPayloadTail> {
        match self {
            PartialPayload::Tail(tail) => Some(tail),
            _ => None,
        }
    }

    pub fn get_head_mut(&mut self) -> Option<&mut PartialPayloadHead> {
        match self {
            PartialPayload::Head(head) => Some(head),
            _ => None,
        }
    }

    pub fn get_tail_mut(&mut self) -> Option<&PartialPayloadTail> {
        match self {
            PartialPayload::Tail(tail) => Some(tail),
            _ => None,
        }
    }
}
