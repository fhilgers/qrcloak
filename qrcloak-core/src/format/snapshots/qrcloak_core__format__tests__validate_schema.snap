---
source: qrcloak-core/src/format/mod.rs
expression: schema
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "Payload",
  "description": "The payload format, being either a complete or partial payload.",
  "anyOf": [
    {
      "description": "A complete payload that is not split accross multiple partial ones.",
      "allOf": [
        {
          "$ref": "#/definitions/CompletePayload"
        }
      ]
    },
    {
      "description": "A partial payload that is split accross multiple partial ones.",
      "allOf": [
        {
          "$ref": "#/definitions/PartialPayload"
        }
      ]
    }
  ],
  "definitions": {
    "CompletePayload": {
      "description": "A complete payload, meaning a payload that is not split accross multiple partial ones.",
      "type": "object",
      "required": [
        "data"
      ],
      "properties": {
        "compression": {
          "description": "The compression to be used for the payload.",
          "allOf": [
            {
              "$ref": "#/definitions/CompressionSpec"
            }
          ]
        },
        "data": {
          "description": "The data of the payload.",
          "type": "string",
          "format": "base45",
          "pattern": "^[0-9A-Z\\s\\$%\\*\\+\\-\\.\\/:]*$"
        },
        "encryption": {
          "description": "The encryption to be used for the payload.",
          "allOf": [
            {
              "$ref": "#/definitions/EncryptionSpec"
            }
          ]
        }
      }
    },
    "CompressionSpec": {
      "description": "The specification of the compression to be used for the payload.",
      "oneOf": [
        {
          "description": "No compression is used.",
          "type": "string",
          "enum": [
            "NoCompression"
          ]
        }
      ]
    },
    "EncryptionSpec": {
      "description": "The specification of the encryption to be used for the payload.",
      "oneOf": [
        {
          "description": "No encryption is used.",
          "type": "string",
          "enum": [
            "NoEncryption"
          ]
        },
        {
          "description": "The payload is encrypted with a passphrase using [`age`].",
          "type": "string",
          "enum": [
            "AgePassphrase"
          ]
        },
        {
          "description": "The payload is encrypted with a x25519 key using [`age`].",
          "type": "string",
          "enum": [
            "AgeKey"
          ]
        }
      ]
    },
    "Index": {
      "description": "The index of a partial payload.",
      "type": "object",
      "required": [
        "id",
        "index",
        "size"
      ],
      "properties": {
        "id": {
          "description": "The id to match multiple partial payloads together",
          "type": "integer",
          "format": "uint32",
          "minimum": 0.0
        },
        "index": {
          "description": "The index in the group",
          "type": "integer",
          "format": "uint32",
          "minimum": 0.0
        },
        "size": {
          "description": "The total size of the group",
          "type": "integer",
          "format": "uint32",
          "minimum": 0.0
        }
      }
    },
    "PartialPayload": {
      "description": "A partial payload",
      "oneOf": [
        {
          "description": "The head of the group of partial payloads.",
          "type": "object",
          "required": [
            "Head"
          ],
          "properties": {
            "Head": {
              "$ref": "#/definitions/PartialPayloadHead"
            }
          },
          "additionalProperties": false
        },
        {
          "description": "The following partial payloads in the group.",
          "type": "object",
          "required": [
            "Tail"
          ],
          "properties": {
            "Tail": {
              "$ref": "#/definitions/PartialPayloadTail"
            }
          },
          "additionalProperties": false
        }
      ]
    },
    "PartialPayloadHead": {
      "description": "A partial payload head, meaning the first partial payload in a group. This payload carries additional information about encryption and compression.",
      "type": "object",
      "required": [
        "data",
        "index"
      ],
      "properties": {
        "compression": {
          "description": "The compression to be used for the group of partial payloads.",
          "allOf": [
            {
              "$ref": "#/definitions/CompressionSpec"
            }
          ]
        },
        "data": {
          "description": "The data of the payload.",
          "type": "string",
          "format": "base45",
          "pattern": "^[0-9A-Z\\s\\$%\\*\\+\\-\\.\\/:]*$"
        },
        "encryption": {
          "description": "The encryption to be used for the group of partial payloads.",
          "allOf": [
            {
              "$ref": "#/definitions/EncryptionSpec"
            }
          ]
        },
        "index": {
          "description": "The index of the payload.",
          "allOf": [
            {
              "$ref": "#/definitions/Index"
            }
          ]
        }
      }
    },
    "PartialPayloadTail": {
      "description": "A partial payload tail, meaning any partial payload that is not the head (the first).",
      "type": "object",
      "required": [
        "data",
        "index"
      ],
      "properties": {
        "data": {
          "description": "The data of the payload.",
          "type": "string",
          "format": "base45",
          "pattern": "^[0-9A-Z\\s\\$%\\*\\+\\-\\.\\/:]*$"
        },
        "index": {
          "description": "The index of the payload.",
          "allOf": [
            {
              "$ref": "#/definitions/Index"
            }
          ]
        }
      }
    }
  }
}
